### ISSUE TITLE
proposal: add a global option for configuring StackFrames

### ISSUE THREAD

**Issue by mrVanboy** [DataDog/dd-trace-go#443]
After rebasing our project to the new version with `StackFrames`, I realized, that it requires changes in many `contrib/*` subpackages too. Our usecase is gRPC middleware for example. 

Maybe it worth time to implement similar functionality as a `tracer.StartOption` to configure frames only once at a single place during tracer initialization? We could prepare another PR for that.

_Originally posted by @mrVanboy in https://github.com/DataDog/dd-trace-go/pull/436#issuecomment-495205371_

**Comment by gbbr** [comment]
I'm not convinced that we want to duplicate this feature there, in `tracer.StartOption`. On the other hand, it definitely seems like a better place to have it, rather than keep repeating it on every finishing span (nevertheless repeating it can be avoided by making your own custom helper).

As I see it, we have two options here:

1. Create a new `tracer.StartOption` call which applies globally to all finishing spans, but keep the current `tracer.FinishOption` that you've added which would override any global setting.
2. Add a new option to the gRPC package which automatically adds the already existing feature to all finished spans.

Do you think having 1. would be a good solution for you? Would you be happy to have one global setting that would apply to _all_ spans of that process?

**Comment by mrVanboy** [comment]
> nevertheless repeating it can be avoided by making your own custom helper)

We have a custom package as a wrapper of the `dd-trace-go`. It solves the problem with setting `FinishOption` only once, but it works only for spans created by hand and unfortunately, it does not cover spans in (probably) all `contrib` packages (`gRPC` in our case).

I definitely agree that both options (`tracer.StartOption` and `tracer.FinishOption`) should exist to make behavior modifications simpler.

1. Sounds good. I don't have a deep knowledge of the `tracer` package, but I think it should work 🙂.
2. Is simpler for solving our particular issue, but I'm not sure how good it could be maintainable because this (ideally) should be implemented in all `contrib` sub-packages.

@matoous what do you think?

**Comment by matoous** [comment]
I agree. Can't say much more. I can work on it again if you want.

**Comment by gbbr** [comment]
If you're open to doing that, we're happy to accept your contributions!

It should likely help with #445 too. The author of that issue has also encountered difficulties because of the way our stack starts inside the tracer, as opposed to inside his application.

**Comment by knusbaum** [comment]
Since #450 was rejected, we should decide how to proceed on this proposal.
@gbbr, is this proposal still viable? Should we attempt option 2 (Add a new option to the gRPC package)?

**Comment by knusbaum** [comment]
Closing this as stale without marking it rejected. This is not something there's a lot of interest in doing right now, but we may revisit it at a later date.
